xcpretty: format additional build steps.

alon-k opened the pull request:
Add formatting for a few more build steps.
@byohay

alon-k committed:
xcpretty: format SymLink.

alon-k committed:
xcpretty: refactor ProcessPCH format to work for ProcessPCH++.

Also remove ProcessPCH command from parser since it isn't used in
simple formatter (it's only in the json reporter) and it has very
bad performance.

alon-k committed:
xcpretty: format swift related build steps.

alon-k committed:
xcpretty: refactor parsing of generic warnings.

The generic warning parsing wasn't generic enough and didn't catch a
a few cases which appeared in our logs. Tests for the above cases were
added and the parser was edited to pass them.

alon-k committed:
xcpretty: add formatting for some archiving steps.

alon-k commented:
Fixed cucumber issue. Forgot to run integration tests and had a small inconsistency between formatter.rb and simple.rb APIs.

byohay commented on spec/xcpretty/formatters/simple_spec.rb:
```diff
@@ -67,6 +67,12 @@ module XCPretty
       end
 
+      it "formats sym link output" do
+        @formatter.format_sym_link('Source.h',
```
Add folder path to source.

alon-k replied:
Done

byohay commented on spec/fixtures/constants.rb:
```diff
@@ -626,6 +626,12 @@
 )
 
+SAMPLE_SYM_LINK = %Q(
+SymLink /Users/jenkins/Library/Developer/Xcode/DerivedData/Blah-fsncgmwwwwqfhufpjvhbxnyuqkxp/Build/Intermediates.noindex/ArchiveIntermediates/Blah/BuildProductsPath/Release-iphoneos/SomeFramework.framework /Users/jenkins/Library/Developer/Xcode/DerivedData/Blah-fsncgmwwwwqfhufpjvhbxnyuqkxp/Build/Intermediates.noindex/ArchiveIntermediates/Blah/IntermediateBuildFilesPath/UninstalledProducts/iphoneo ...
```
Better to use a different name for the target so that the test will be more precise.

alon-k replied:
Done

byohay commented on lib/xcpretty/parser.rb:
```diff
@@ -167,12 +167,9 @@ module Matchers
-    # $1 file_path
-    PROCESS_PCH_COMMAND_MATCHER = /^\s*.*\/usr\/bin\/clang\s.*\s\-c\s(.*)\s\-o\s.*/.freeze
+    # $1 = nil or ++ (C or C++)
+    # $2 = file
```
`file_path`

alon-k replied:
Fixed

byohay commented on lib/xcpretty/parser.rb:
```diff
@@ -167,12 +167,9 @@ module Matchers
-    PROCESS_PCH_COMMAND_MATCHER = /^\s*.*\/usr\/bin\/clang\s.*\s\-c\s(.*)\s\-o\s.*/.freeze
+    # $1 = nil or ++ (C or C++)
+    # $2 = file
+    PROCESS_PCH_MATCHER = /^ProcessPCH(\+\+)?\s.*\s(.*.pch)/.freeze
```
For literal dot use `\.`.

alon-k replied:
Fixed

byohay commented on lib/xcpretty/formatters/formatter.rb:
```diff
@@ -38,7 +38,7 @@ def format_passing_test(suite, test, time);                EMPTY; end
     def format_measuring_test(suite, test, time);              EMPTY; end
     def format_failing_test(suite, test, reason, file_path);   EMPTY; end
-    def format_process_pch(file, build_target);                EMPTY; end
+    def format_process_pch(cpp, file, build_target);           EMPTY; end
```
I prefer that the formatter will get the language and not just the postfix, that is, the parser will add the `C`. Then the interface between the parser and the formatter are nicer.

alon-k replied:
Changed

byohay commented on lib/xcpretty/formatters/formatter.rb:
```diff
@@ -38,7 +38,7 @@ def format_passing_test(suite, test, time);                EMPTY; end
     def format_failing_test(suite, test, reason, file_path);   EMPTY; end
-    def format_process_pch(file, build_target);                EMPTY; end
+    def format_process_pch(cpp, file, build_target);           EMPTY; end
     def format_process_pch_command(file_path, build_target);   EMPTY; end
```
It exists in the JSON formatter, this is where it is used. The question is whether we completely disregard the JSON formatter for now, or we should try to break it as less as possible. @barakwei?

alon-k replied:
I'll put it back for now, but would also like to know what @barakwei thinks about this

alon-k replied:
Also, if we are talking performance: the regex that matches this is one of the worst ones in terms of performance (my 80k step regex before we fixed it was based on it), which means that even though we don't use it, since we left it in it causes a hit to performance. It might be a good idea to remove it after all. Might also be a good idea to go over the other regexes and optimize them, and also remove everything we don't use, since the parser still attempts to match even the regexes that we don't use for simple formatter and removing them might also improve performance.

alon-k replied:
Spoke to Barak and removed it.

byohay commented on lib/xcpretty/parser.rb:
```diff
@@ -64,6 +64,23 @@ module Matchers
+    SWIFT_CODE_GENERATION_MATCHER = /^SwiftCodeGeneration\s.*$/.freeze
+
+    # @regex Captured groups
+    # $1 = bytecode file
```
`bytecode file path`

alon-k replied:
Fixed

byohay commented on lib/xcpretty/parser.rb:
```diff
@@ -64,6 +64,23 @@ module Matchers
+
+    # @regex Captured groups
+    # $1 = bytecode file
+    SWIFT_CODE_GENERATION_COMMAND_MATCHER = /^\s*.*\/usr\/bin\/swift.*\s-c\s.*\s(.*\.bc)/.freeze
```
1. @barakwei this is not the primary command, do we really need this?
2. Regex can be dangerously slow! For example the sample code takes 80k steps to match: https://regex101.com/r/6Anrlz/1. We already had performance issues with xcpretty, so we want to keep it fast.
There are several ways to do it: In this case, using `.*` or anything similar is expensive, you should use avoid it when you can. For example: You don't need `^\s*.*` at the beginning, and I'm not sure that you need to find `-c` also, in which case you can merge the two `.*`. Please check your other regexes as well.

barakwei replied:
I don't think we need it, but @alon-k insisted, so I leave the decision to you.

alon-k replied:
1. I think that either this should be included, or xcpretty shouldn't print anything at all for SwiftCodeGeneration logs, as otherwise they contain no meaningful information other than the target and the number of object files being generated (notice that the file path is only included in the command, and not in the SwiftCodeGeneration log - btw, @barakwei didn't you want to open an issue about that?)
2. I had to find -c because otherwise it matches other /usr/bin/swift commands from earlier in the log (in the other swift build commands) which are unwanted as they compile a bunch of files at the same time and are just hard to work with in general.
Regarding the `^\s*.*` in the beginning, I've removed it. This cut down most of the steps (only takes 5k steps to match now), so I think the performance hit of the -c isn't very significant.

alon-k replied:
edit: I checked and the -c barely costs any steps anyways. Removing it brings it down from 5k to 4.6k steps at the cost of matching things we don't want to match

alon-k replied:
improved further (to around 1k steps) in latest push.

alon-k replied:
improved to around 400 steps in latest push

byohay commented on lib/xcpretty/formatters/simple.rb:
```diff
@@ -44,6 +44,27 @@ def format_compile(file_name, file_path, build_target)
+    end
+
+    def format_copy_swift_libs(app_path, build_target)
+      format("Copying required Swift libs for", File.basename(app_path),
```
Why required? Aren't all copied libs required?

alon-k replied:
They are. This says that the libs which are copied (and thus **required**) are being copied right now. The wording can be changed if you'd like.

alon-k replied:
Fixed.

alon-k replied:
Changed in latest push.

byohay commented on lib/xcpretty/parser.rb:
```diff
@@ -46,6 +46,24 @@ module Matchers
     CLEAN_TARGET_MATCHER = /^=== CLEAN TARGET\s(.*)\sOF PROJECT\s(.*)\sWITH CONFIGURATION\s(.*)\s===/.freeze
 
+    # @regex Captured groups
+    # $1 = app file
```
Specify weather this is just the name or path: `app file path/name`. Same for other places.

alon-k replied:
Done

byohay commented on lib/xcpretty/parser.rb:
```diff
@@ -303,7 +303,7 @@ module Warnings
       # @regex Captured groups
       # $1 = whole warning
-      GENERIC_WARNING_MATCHER = /^warning:\s(.*)$/.freeze
+      GENERIC_WARNING_MATCHER = /^.*[Ww]arning:\s(.*)$/.freeze
```
You can simplify by doing: `[Ww]arning:\s(.*)$`

alon-k replied:
Notice that this will make the libpng test case fail as it will cause the regex to capture "Warning: blah blah" instead of "blah blah" in that case. It does make the regex faster though. We can let it slide for the sake of performance, or leave it as it is - your call.

alon-k replied:
Left as is for now. The performance hit isn't very significant here either - 100 steps instead of 30 steps.

byohay reviewed (changes requested):
Round.

byohay commented:
@alon-k also fix PR title.

alon-k commented:
Round complete.
PR Title fixed.

alon-k commented:
Pushed some more fixes.

byohay commented on lib/xcpretty/formatters/formatter.rb:
```diff
@@ -38,7 +38,7 @@ def format_passing_test(suite, test, time);                EMPTY; end
     def format_measuring_test(suite, test, time);              EMPTY; end
     def format_failing_test(suite, test, reason, file_path);   EMPTY; end
-    def format_process_pch(file, build_target);                EMPTY; end
+    def format_process_pch(cpp, file, build_target);           EMPTY; end
```
Change to `language`.

alon-k replied:
Done

byohay commented on lib/xcpretty/parser.rb:
```diff
@@ -64,6 +82,23 @@ module Matchers
+    COMPILE_SWIFT_MATCHER = /^CompileSwift\s.*$/.freeze
+
+    # @regex Captured groups
+    SWIFT_CODE_GENERATION_MATCHER = /^SwiftCodeGeneration\s.*$/.freeze
```
Consider using state of build_target, then you can get rid of this regex altogether.

alon-k replied:
Done

byohay reviewed (changes requested):
F&S.

alon-k commented:
Pushing last fix + additional swift related formatting that was discovered while testing other projects. Please wait before reviewing

alon-k committed:
parser: fix parsing process product packaging.

The regex that was used before was too specific (it only matched
entitlements, whereas sometimes this log also appeared with provisioning
profiles, and also only matched if they were wrapped in `"`, which
wasn't always the case). It also didn't properly work with escaped
spaces in the paths. Tests were added for the above cases and the
parser was edited to pass them.

alon-k commented:
Done. Please review.

byohay commented on spec/xcpretty/parser_spec.rb:
```diff
@@ -456,13 +456,22 @@ module XCPretty
       @parser.parse(SAMPLE_PROCESS_PRODUCT_PACKAGING)
     end
 
+    it "parses process product packaging #2" do
```
`parses process product packaging without entitlements`

alon-k replied:
Fixed

byohay commented on lib/xcpretty/parser.rb:
```diff
@@ -275,9 +275,9 @@ module Matchers
+    # $1 = entitlements / provisioning + directory they are in
     # $2 = package name
-    PROCESS_PRODUCT_PACKAGING_MATCHER = /^ProcessProductPackaging\s"((?:\\"|[^"])*)"\s.+\/(.+)$/.freeze
+    PROCESS_PRODUCT_PACKAGING_MATCHER = /^ProcessProductPackaging\s((?:\\\s|\S)*)\s.+\/(.+\/.+)$/.freeze
```
Put `"` as optional.

alon-k replied:
Done

byohay commented on lib/xcpretty/parser.rb:
```diff
@@ -275,9 +275,9 @@ module Matchers
+    # $1 = entitlements / provisioning + directory they are in
     # $2 = package name
-    PROCESS_PRODUCT_PACKAGING_MATCHER = /^ProcessProductPackaging\s"((?:\\"|[^"])*)"\s.+\/(.+)$/.freeze
+    PROCESS_PRODUCT_PACKAGING_MATCHER = /^ProcessProductPackaging\s((?:\\\s|\S)*)\s.+\/(.+\/.+)$/.freeze
```
Why did you change `.+` to `.+\/.+`?

alon-k replied:
Changed back

byohay commented on lib/xcpretty/parser.rb:
```diff
@@ -275,9 +275,9 @@ module Matchers
 
     # @regex Captured groups
-    # $1 = entitlements
+    # $1 = entitlements / provisioning + directory they are in
```
Unclear what you mean.

alon-k replied:
Fixed

byohay reviewed (changes requested):
F&S.

alon-k commented:
Done.

byohay commented on lib/xcpretty/parser.rb:
```diff
@@ -271,9 +271,9 @@ module Matchers
 
     # @regex Captured groups
-    # $1 = entitlements
+    # $1 = entitlements / provisioning profile name
```
name -> path

alon-k replied:
Fixed

byohay reviewed (changes requested):
F&S.

alon-k commented:
Done.
